### PR TITLE
Fixed #34077 -- Added form field rendering.

### PR DESCRIPTION
**Pull request by smithdc1**
The first commit adjusts the test added for [ticket #33134](https://code.djangoproject.com/ticket/33134). The ticket opened was all about django-debug-toolbar recursively calling `str()` on a template's `context`. However, the test I wrote while it demonstrated the cyclical issue it used `render()` not `str()`. 

The next commit makes changes such that there'll be a recursion issue if calling `render()` but not `str()` (we're making no changes to boundfiled's `str()` function as part of this commit. 

I _think_ that's all ok as far as not breaking DDT is concerned. 

The second commit here is step 1 as discussed at #34077 and is the first two commits from there merged, and I've added some more docs to the topic section. I wonder if there's more documentation updates that should be made as part of this patch.

### CODE REVIEW

**Comment by carltongibson** [comment]
Hey @smithdc1 — I'm going to merge the first commit. DDT is tracking https://github.com/jazzband/django-debug-toolbar/issues/1509 — which would use `str()` evening if reverting [DDT/1a3cab](https://github.com/jazzband/django-debug-toolbar/commit/1a3cabac1c3d9b98b0d024a1bb907f20f487c5cf) — so, agreed, us checking that does not affect them.

**Review by carltongibson — APPROVED** [review]
Hey @smithdc1. 

I pushed some suggested docs edits. Please let me know what you think. 

tl;dr I really like this. Let me ask @felixxm to look, but I think we should be able to progress. (And then we have the rest of the cycle to consider if/when/how for the step 2 deprecation bit.) 

I'm going to put a ✅ :)

**Comment by carltongibson** [comment]
Not sure about https://github.com/django/django/pull/16666/commits/51948995e1a68b4784a05791363930e8ba0ed220 — Should the spell check pick up code blocks? 🤔

**Comment by smithdc1** [comment]
> Please let me know what you think.

They look good to me 👍. Thank you for the review. 

>  Should the spell check pick up code blocks? 🤔

I'm not sure, it seems odd to me. Side note, I can not get `PyEnchant` to work on Windows (something to do with a dll needing to be rebuilt 🤷 ), further `PyEnchant` is looking for maintainers.

**Comment by carltongibson** [comment]
It may be that `codespell` could be an alternative 

https://til.simonwillison.net/python/codespell

File: docs/topics/forms/index.txt [hunk 1]
```diff
@@ -559,13 +559,73 @@ the :meth:`.Form.render`. Here's an example of this being used in a view::
...
+    <div class="fieldWrapper">
+    {{ form.subject.as_field_group }}
+    </div>
+    <div class="fieldWrapper">
+    {{ form.message.as_field_group }}
+    </div>
+    <div class="fieldWrapper">
+    {{ form.sender.as_field_group }}
+    </div>
+    <div class="fieldWrapper">
```
Comment (felixxm) [inline docs/topics/forms/index.txt:590]: Can we use `field-wrapper` to avoid adding code words to the `spelling_wordlist`?

File: docs/topics/forms/index.txt [hunk 2]
```diff
@@ -559,13 +559,73 @@ the :meth:`.Form.render`. Here's an example of this being used in a view::
...
+.. versionadded:: 5.0
+
+Each field is available as an attribute of the form, using ``{{
+form.name_of_field }}`` in a template. A field has a
+:meth:`~django.forms.BoundField.as_field_group` method which renders the
+related elements of the field as a group, its label, widget, errors and help
+text.
+
+This allows generic templates to be written that arrange fields elements in the
+required layout. For example::
```
Comment (felixxm) [inline docs/topics/forms/index.txt:576]:
```suggestion
required layout. For example:
```

File: docs/topics/forms/index.txt [hunk 3]
```diff
@@ -559,13 +559,73 @@ the :meth:`.Form.render`. Here's an example of this being used in a view::
...
+
+The default template can be customized by by setting
+:attr:`~django.forms.renderers.BaseRenderer.field_template_name` in your
+project-level :setting:`FORM_RENDERER`::
+
+    from django.forms.renderers import TemplatesSetting
+
+
+    class CustomFormRenderer(TemplatesSetting):
+        form_template_name = "form_snippet.html"
```
Comment (felixxm) [inline docs/topics/forms/index.txt:605]:
```suggestion
        field_template_name = "field_snippet.html"
```

File: docs/topics/forms/index.txt [hunk 4]
```diff
@@ -559,13 +559,73 @@ the :meth:`.Form.render`. Here's an example of this being used in a view::
...
+
+… or on a single field::
+
+    class MyForm(forms.Form):
+        subject = forms.CharField(template_name="my_custom_template.html")
+        ...
+
+… or on a per-request basis by calling
+  :meth:`~django.forms.BoundField.render` and supplying a template name::
```
Comment (felixxm) [inline docs/topics/forms/index.txt:614]:
```suggestion
:meth:`~django.forms.BoundField.render` and supplying a template name::
```

File: docs/topics/forms/index.txt [hunk 5]
```diff
@@ -559,13 +559,73 @@ the :meth:`.Form.render`. Here's an example of this being used in a view::
 
 See :ref:`ref-forms-api-outputting-html` for more details.
 
+.. _reusable-field--group-templates:
```
Comment (felixxm) [inline docs/topics/forms/index.txt:562]:
```suggestion
.. _reusable-field-group-templates:
```

**Review by felixxm — COMMENTED** [review]
@smithdc1 Thanks :+1: Great work :rocket: 

I added tests for `NotImplementedError`, _"Overriding built-in field templates"_ section, and pushed small edits to docs.

File: docs/topics/forms/index.txt (hunk 2 above)
Comment (felixxm) [reply docs/topics/forms/index.txt]: :point_up: This is the reason the `spelling` complained about `fieldWrapper` :smile:

**Comment by felixxm** [comment]
buildbot, test on selenium.
